### PR TITLE
Adding Placeholder API Compability internally [WIP]

### PR DESCRIPTION
**Pull request by Banbeucmas**
Pretty sure I want this to be an enhancement in 2.16.0
Still WIP tho.

Just that the placeholder seem to work flawlessly with PlaceholderAPI 2.8.5 tested

image

image

I hope my pr didn't fucked up lol. Haven't try to merge specific branch yet

### CODE REVIEW

**Comment by Banbeucmas** [comment]
Yep it fucked up

**Comment by Banbeucmas** [comment]
A little note:
image

**Comment by Banbeucmas** [comment]
As for registering our own placeholder, https://github.com/PlaceholderAPI/Essentials-Expansion is already here, I think it is better to just continue to contribute there rather than putting them internally onto Essentials

**Comment by Banbeucmas** [comment]
[https://i.imgur.com/lGFNgdV_d.jpg?maxwidth=640&shape=thumb&fidelity=medium](url)

**Comment by Banbeucmas** [comment]
https://ibb.co/hn78wc

**Comment by Banbeucmas** [comment]
Nice one.

File: Essentials/src/plugin.yml [hunk 1]
```diff
@@ -5,7 +5,7 @@ main: com.earth2me.essentials.Essentials
 version: ${full.version}
 website: http://tiny.cc/EssentialsCommands
 description: Provides an essential, core set of commands for Bukkit.
-softdepend: [Vault]
+softdepend: [Vault, PlaceholderAPI]
```
Comment (Banbeucmas) [inline Essentials/src/plugin.yml:8]: https://github.com/EssentialsX/Essentials/issues/1945
You once said you don't want to add it as sofedepend.
Will we remove this?

**Review by Banbeucmas — COMMENTED** [review]
plugin.yml might need some work

File: Essentials/src/com/earth2me/essentials/utils/FormatUtil.java [hunk 2]
```diff
@@ -38,6 +43,30 @@ public static String stripEssentialsFormat(final String input) {
...
+    }
+
+    //Formatting PlaceholderAPI Placeholders, returns the original string if the plugin isn't enabled
+    //Seperate this from #getChatFormat() in ISettings due to the requirement of Player as an argument
+    public static String placeholderAPIFormat(final Player player, String input) {
+        if (input == null) {
+            return null;
+        }
+
+        input = input.replaceAll("\\{|\\}", "%");
```
Comment (Banbeucmas) [inline Essentials/src/com/earth2me/essentials/utils/FormatUtil.java:58]: There might be some problem that a single { standing alone will become % after that. I don't use regex for a long time. Should we do something about it.

**Review by Banbeucmas — COMMENTED** [review]
Was aware of this tho

File: Essentials/src/com/earth2me/essentials/utils/FormatUtil.java (hunk 2 above)
Comment (mdcfe) [reply Essentials/src/com/earth2me/essentials/utils/FormatUtil.java:58]: `/{([A-Za-z0-9_]+)}/g` might work.

File: Essentials/src/plugin.yml (hunk 1 above)
Comment (mdcfe) [reply Essentials/src/plugin.yml:8]: Oops, forgot about that. Still need to implement handling of PAPI loading after EssentialsX though before we do that.

Comment (Banbeucmas) [reply Essentials/src/plugin.yml:8]: I will leave that part to you then :)

File: Essentials/src/com/earth2me/essentials/utils/FormatUtil.java (hunk 2 above)
Comment (Banbeucmas) [reply Essentials/src/com/earth2me/essentials/utils/FormatUtil.java:58]: @md678685 I love IntelliJ
image

File: Essentials/src/com/earth2me/essentials/utils/FormatUtil.java [hunk 3]
```diff
@@ -54,11 +55,7 @@ public static String placeholderAPIFormat(final Player player, String input) {
             return null;
         }
 
-        input = input.replaceAll("\\{|\\}", "%");
-
-        if (papiEnabled == null) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/utils/FormatUtil.java]: This still needs to be here, in case PAPI loads before EssentialsX - in that instance, Essentials won't receive PAPI's `PluginEnableEvent`.

Comment (mdcfe) [reply Essentials/src/com/earth2me/essentials/utils/FormatUtil.java]: That is also the reason for the null - at the point that the class is loaded, we won't know whether PAPI is enabled, and it also might not be a safe time to check.

Comment (Banbeucmas) [reply Essentials/src/com/earth2me/essentials/utils/FormatUtil.java]: Ah, that make sense.  Let me change that

Comment (Banbeucmas) [reply Essentials/src/com/earth2me/essentials/utils/FormatUtil.java]: Tbh, we can put false rather than null. If PluginEvent didn't fired while PAPI excistance still there then it would still false anyway. `false` seem more appropiate this time.

File: Essentials/src/com/earth2me/essentials/utils/FormatUtil.java (hunk 2 above)
Comment (Banbeucmas) [reply Essentials/src/com/earth2me/essentials/utils/FormatUtil.java:58]: Still need to address this.

Comment (mdcfe) [reply Essentials/src/com/earth2me/essentials/utils/FormatUtil.java:58]: I think the replacement with that regex would have been `"%$1%"`, since the `()`s in the regex capture the placeholder name (`bobsplugin_hunger`) and the `$1` puts it back in the string once replaced.

Comment (Banbeucmas) [reply Essentials/src/com/earth2me/essentials/utils/FormatUtil.java:58]: I love how the problem still excist tho. Idk why but seem like Java Regex behaves different than JS

File: Essentials/src/com/earth2me/essentials/Teleport.java [hunk 4]
```diff
@@ -247,6 +249,13 @@ protected void respawnNow(IUser teleportee, TeleportCause cause) throws Exceptio
     //The warp function is a wrapper used to teleportPlayer a player to a /warp
     @Override
     public void warp(IUser teleportee, String warp, Trade chargeFor, TeleportCause cause) throws Exception {
+        UserWarpEvent event = new UserWarpEvent(teleportee, warp, chargeFor);
```
Comment (drtshock) [inline Essentials/src/com/earth2me/essentials/Teleport.java:252]: should be in its own PR?

Comment (mdcfe) [reply Essentials/src/com/earth2me/essentials/Teleport.java:252]: Not sure why this change is in this PR but the actual PR is [here](https://github.com/EssentialsX/Essentials/pull/1921).

**Comment by Banbeucmas** [comment]
Yeah, I accidentally create the Papi branch out of the original branch of my fork and PR before even realize it.
We will just solve both PR at the same time or if anyone have a better idea.

File: Essentials/src/com/earth2me/essentials/Teleport.java (hunk 4 above)
Comment (Banbeucmas) [reply Essentials/src/com/earth2me/essentials/Teleport.java:252]: Yeah, I accidentally create the Papi branch out of the original branch of my fork and PR before even realize it.
We will just solve both PR at the same time or if anyone have a better idea.

**Comment by Banbeucmas** [comment]
Should we close this in favor of https://github.com/EssentialsX/Essentials/issues/2041?

**Comment by mdcfe** [comment]
@Banbeucmas Yes, closing this PR now.
